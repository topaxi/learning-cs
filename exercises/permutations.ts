export function permutations(
  nums: number[],
  added = new Set<number>(),
  list: number[] = [],
  out: number[][] = []
): number[][] {
  if (nums.length === list.length) return out.push(list.slice()), out

  for (let i = 0; i < nums.length; i++) {
    let n = nums[i]
    if (!added.has(n)) {
      added.add(n)
      list.push(n)
      permutations(nums, added, list, out)
      list.pop()
      added.delete(n)
    }
  }

  return out
}
