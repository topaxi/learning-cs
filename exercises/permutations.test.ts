import { permutations } from './permutations'

describe('46. Permutations', () => {
  test('should generate each permutation', () => {
    expect(permutations([1, 2, 3])).toMatchSnapshot()
    expect(permutations([1, 2, 3, 4])).toMatchSnapshot()
  })
})
