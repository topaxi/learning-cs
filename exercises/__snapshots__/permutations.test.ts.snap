// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`46. Permutations should generate each permutation 1`] = `
[
  [
    1,
    2,
    3,
  ],
  [
    1,
    3,
    2,
  ],
  [
    2,
    1,
    3,
  ],
  [
    2,
    3,
    1,
  ],
  [
    3,
    1,
    2,
  ],
  [
    3,
    2,
    1,
  ],
]
`;

exports[`46. Permutations should generate each permutation 2`] = `
[
  [
    1,
    2,
    3,
    4,
  ],
  [
    1,
    2,
    4,
    3,
  ],
  [
    1,
    3,
    2,
    4,
  ],
  [
    1,
    3,
    4,
    2,
  ],
  [
    1,
    4,
    2,
    3,
  ],
  [
    1,
    4,
    3,
    2,
  ],
  [
    2,
    1,
    3,
    4,
  ],
  [
    2,
    1,
    4,
    3,
  ],
  [
    2,
    3,
    1,
    4,
  ],
  [
    2,
    3,
    4,
    1,
  ],
  [
    2,
    4,
    1,
    3,
  ],
  [
    2,
    4,
    3,
    1,
  ],
  [
    3,
    1,
    2,
    4,
  ],
  [
    3,
    1,
    4,
    2,
  ],
  [
    3,
    2,
    1,
    4,
  ],
  [
    3,
    2,
    4,
    1,
  ],
  [
    3,
    4,
    1,
    2,
  ],
  [
    3,
    4,
    2,
    1,
  ],
  [
    4,
    1,
    2,
    3,
  ],
  [
    4,
    1,
    3,
    2,
  ],
  [
    4,
    2,
    1,
    3,
  ],
  [
    4,
    2,
    3,
    1,
  ],
  [
    4,
    3,
    1,
    2,
  ],
  [
    4,
    3,
    2,
    1,
  ],
]
`;
